File: src/openagents/tool/skill_tool.py
import json
import logging
import pathlib

import anyio
from pydantic import BaseModel


# Skill数据结构
class SkillInfo(BaseModel):
    name: str
    description: str
    path: str
    content: str


# Skill读取目录
SKILLS_DIR_LIST = [str(pathlib.Path.home() / ".openagents" / "skills"), str(pathlib.Path.home() / ".agents" / "skills")]
# 存储所有的skill信息
SKILLS: list[SkillInfo] = []


# 初始化所有skill信息
async def init_skills() -> None:
    loaded = set()
    # 遍历技能目录
    for skills_dir in map(anyio.Path, SKILLS_DIR_LIST):
        if not await skills_dir.exists():
            continue
        # 遍历技能
        async for skill_dir in skills_dir.iterdir():
            if skill_dir.name in loaded:
                continue
            skill_file = skill_dir / "SKILL.md"
            if not await skill_file.is_file():
                continue
            # 尝试读取 SKILL.md 提取 name 和 description
            text = await skill_file.read_text(encoding="utf-8")
            lines = [line.strip() for line in text.split("\n")]
            if lines[0] == "---" and "---" in lines[1:]:
                second_index = lines.index("---", 1)
            else:
                continue
            name, description = "", ""
            for line in lines[1:second_index]:
                if line.startswith("name:"):
                    name = line[5:].strip()
                elif line.startswith("description:"):
                    description = line[12:].strip()
            if name == skill_dir.name:
                content = text.split("---\n", 2)[2].strip()
                SKILLS.append(SkillInfo(name=name, description=description, path=str(await skill_file.resolve()), content=content))
                loaded.add(name)
    logging.info(f"Skills initialized, having {len(SKILLS)} skills")


def list_skills() -> list[SkillInfo]:
    return SKILLS


# 执行
async def execute(cmd_and_args: list[str], work_dir: str) -> tuple[str, bool]:
    # 1. skill list
    if len(cmd_and_args) == 2 and cmd_and_args[0] == "skill" and cmd_and_args[1] == "list":
        tool_content = json.dumps([{"name": s.name, "description": s.description, "path": s.path} for s in SKILLS], ensure_ascii=False)
        return tool_content, False
    return "Unknown command", True

File: src/tool/skill_tool.rs
// 技能列表工具
use serde::Serialize;
use std::path::PathBuf;

use super::ToolResult;

// Skill 数据结构
#[derive(Debug, Clone, Serialize)]
pub struct SkillInfo {
    pub name: String,
    pub description: String,
    pub path: String,
    pub content: String,
}

// 存储所有的 skill 信息
static SKILLS: std::sync::RwLock<Vec<SkillInfo>> = std::sync::RwLock::new(Vec::new());

// Skill 读取目录
fn skills_dir_list() -> Vec<PathBuf> {
    let home = std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .unwrap_or_else(|_| ".".to_string());
    vec![
        PathBuf::from(&home).join(".openagents").join("skills"),
        PathBuf::from(&home).join(".agents").join("skills"),
    ]
}

// 初始化所有 skill 信息
pub async fn init_skills() {
    let mut loaded = std::collections::HashSet::new();
    let mut skills = Vec::new();

    for skills_dir in skills_dir_list() {
        if !skills_dir.exists() {
            continue;
        }

        let mut entries = match tokio::fs::read_dir(&skills_dir).await {
            Ok(e) => e,
            Err(_) => continue,
        };

        while let Ok(Some(entry)) = entries.next_entry().await {
            let skill_dir = entry.path();
            let skill_name = match skill_dir.file_name().and_then(|n| n.to_str()) {
                Some(n) => n.to_string(),
                None => continue,
            };

            if loaded.contains(&skill_name) {
                continue;
            }

            let skill_file = skill_dir.join("SKILL.md");
            if !skill_file.is_file() {
                continue;
            }

            // 读取 SKILL.md 提取 name 和 description
            let text = match tokio::fs::read_to_string(&skill_file).await {
                Ok(t) => t,
                Err(_) => continue,
            };

            let lines: Vec<&str> = text.lines().map(|l| l.trim()).collect();
            if lines.is_empty() || lines[0] != "---" {
                continue;
            }

            // 找到第二个 ---
            let second_index = match lines[1..].iter().position(|l| *l == "---") {
                Some(i) => i + 1,
                None => continue,
            };

            let mut name = String::new();
            let mut description = String::new();
            for line in &lines[1..second_index] {
                if line.starts_with("name:") {
                    name = line[5..].trim().to_string();
                } else if line.starts_with("description:") {
                    description = line[12..].trim().to_string();
                }
            }

            if name == skill_name {
                // 提取内容(第二个 --- 之后的部分)
                let content_start = text.match_indices("---").nth(1);
                let content = match content_start {
                    Some((idx, _)) => text[idx + 3..].trim().to_string(),
                    None => String::new(),
                };

                let path = skill_file.canonicalize().unwrap_or(skill_file).to_string_lossy().to_string();
                skills.push(SkillInfo {
                    name,
                    description,
                    path,
                    content,
                });
                loaded.insert(skill_name);
            }
        }
    }

    // 更新全局存储
    {
        let mut guard = SKILLS.write().unwrap();
        *guard = skills;
    }

    let count = SKILLS.read().unwrap().len();
    tracing::info!("Skills initialized, having {} skills", count);
}

// 获取技能列表
pub fn list_skills() -> Vec<SkillInfo> {
    SKILLS.read().unwrap().clone()
}

// 执行
pub fn execute(cmd_and_args: &[String]) -> ToolResult {
    // 1. skill list
    if cmd_and_args.len() == 2 && cmd_and_args[0] == "skill" && cmd_and_args[1] == "list" {
        let skills = SKILLS.read().unwrap();
        let result: Vec<serde_json::Value> = skills
            .iter()
            .map(|s| {
                serde_json::json!({
                    "name": s.name,
                    "description": s.description,
                    "path": s.path
                })
            })
            .collect();
        (serde_json::to_string(&result).unwrap_or_default(), false)
    } else {
        ("Unknown command".to_string(), true)
    }
}